Support for multiple ssids.

Timmmeeeh opened the pull request:
This will close #250 
Support for mutiple ssids by displaying a dialog allowing the users to add and delete multiple ssids.

![Screenshot_20200301-144501](https://user-images.githubusercontent.com/60823285/75626878-7bb26900-5bcb-11ea-9d62-7d71bccbe044.jpg)
![Screenshot_20200301-144527](https://user-images.githubusercontent.com/60823285/75626882-7f45f000-5bcb-11ea-8ea9-f5d9562291a2.jpg)

JBassett commented on data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt:
```diff
@@ -21,7 +21,8 @@ class UrlRepositoryImpl @Inject constructor(
         private const val PREF_WEBHOOK_ID = "webhook_id"
         private const val PREF_LOCAL_URL = "local_url"
-        private const val PREF_WIFI_SSID = "wifi_ssid"
+        private const val PREF_WIFI_SSID = "wifi_ssid" // Keep for backwards compatibility
```
Rather than keep it around in live code we should try to use the https://github.com/home-assistant/home-assistant-android/blob/master/common/src/main/java/io/homeassistant/companion/android/common/migrations/Migrations.kt.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/settings/ssid/SsidSummaryProvider.kt:
```diff
@@ -0,0 +1,20 @@
+) : Preference.SummaryProvider<Preference> {
+
+    override fun provideSummary(preference: Preference): CharSequence {
+        val ssids = runBlocking { urlUseCase.getHomeWifiSsids() }
```
This should be getting the values from the preference, not calling from the `UrlUseCase`.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/settings/ssid/SsidSummaryProvider.kt:
```diff
@@ -0,0 +1,20 @@
+
+class SsidSummaryProvider @Inject constructor(
+    private val urlUseCase: UrlUseCase
+) : Preference.SummaryProvider<Preference> {
```
You should be able to reference your actual preference type, not just the generic preference.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/settings/ssid/SsidPreference.kt:
```diff
@@ -0,0 +1,82 @@
+        private const val NO_MAX_LINES = -1
+    }
+
+    @Inject lateinit var ssidSummaryProvider: SsidSummaryProvider
```
You shouldn't need to inject this.  You should be able to create it here.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/PresenterModule.kt:
```diff
@@ -35,6 +37,8 @@ class PresenterModule {
     private lateinit var settingsView: SettingsView
     private lateinit var webView: WebView
+    private lateinit var ssidPreference: SsidPreference
+    private lateinit var ssidDialog: SsidDialogFragment
```
These shouldn't be loaded with dagger.  They should be treated like any of the other preferences and created when the settings screen is loaded.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/settings/SettingsFragment.kt:
```diff
@@ -81,6 +89,21 @@ class SettingsFragment : PreferenceFragmentCompat(), SettingsView {
         SensorWorker.start(context!!)
     }
 
+    override fun onDisplayPreferenceDialog(preference: Preference) {
```
Once the `SsidDialogFragment` works like other preferences this shouldn't be needed.

Timmmeeeh replied:
Implemented the requested changes, for this one I changed it to the official supported way by implementing the `OnPreferenceDisplayDialogCallback` interface in the `SettingsActivity`.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/settings/ssid/SsidDialogFragment.kt:
```diff
@@ -0,0 +1,97 @@
+            .appComponent((requireContext().applicationContext as GraphComponentAccessor).appComponent)
+            .presenterModule(PresenterModule(this))
+            .build()
+            .inject(this)
```
The dialog fragment shouldn't be using dagger to load anything.  It should all be passed into the preference.

Timmmeeeh commented:
Implemented the requested changes, for displaying the dialog, I changed it to the official supported way by implementing the OnPreferenceDisplayDialogCallback interface in the SettingsActivity.

xaviml commented:
Thanks for this PR @Timmmeeeh. I will finally be able to get rid off the DNS server I had to use to overcome the problem of not having support of NAT loopback in my router :D

scop commented:
Ooh, good stuff. When can we expect a release with this change included? ;)

dshokouhi commented:
@scop join the beta :) its already there but there is 1 open issue #498 to keep in mind while adding multiple

scop commented:
Hm, how do I do that? Didn't find anything beta related in the app configuration, nor by looking at the app entry in Google Play, nor searching for "home assistant beta" there.

dshokouhi commented:
try the beta link from here: https://github.com/home-assistant/home-assistant-android#testing-dev-releases

scop commented:
Nice. got it.

DarwinData commented:
Any idea when this new feature will be available on the prod version?
thanks
